### PR TITLE
closure not required for all ProcPointers

### PR DESCRIPTION
**Pull request by stakach**
see crystal-lang/crystal#11711

@asterite came up with a good way to spec the code

```crystal
def foo
  1
end

puts (->foo).pointer.address
puts (->foo).pointer.address

# should we also test `fun`?
fun ffoo
  1
end

puts (->ffoo).pointer.address
puts (->ffoo).pointer.address
```

where you should always get the same value but not actually sure where this spec should live

### CODE REVIEW

**Review by asterite — APPROVED** [review]
Thank you!

**Comment by stakach** [comment]
I have a feeling that we should be waiting for #11763 to be implemented and add that additional check

File: src/compiler/crystal/semantic/main_visitor.cr [hunk 1]
```diff
@@ -1219,11 +1219,19 @@ module Crystal
         end)
       end
 
+      args = node.args
+
+      # If it's something like `->foo(...)` then we also turn it into a closure
+      # because it could be doing a mutlidispatch and that's not supported in ProcPointer.
+      # However, if it's `->foo` (no args) then it can never be a multidispatch.
+      if !obj && !args.empty?
```
[the remark this answers is not in the document]
Comment (stakach) [reply src/compiler/crystal/semantic/main_visitor.cr:1227]: You are probably be right

**Comment by HertzDevil** [comment]
Merely using a top-level call is not enough, because multi-dispatch may still happen between a fun and a def (#4761).

**Comment by asterite** [comment]
Hm... we should probably not allow having a `def` and a `fun` with the same name at the top level.

**Comment by straight-shoota** [comment]
I'm closing this per https://github.com/crystal-lang/crystal/issues/11711#issuecomment-1018934797
This is clearly not the way to move forward. Discussion can continue in #11711

**Pull request closed without merging**